dbms/src/Storages/DeltaMerge/Segment.cpp

```diff
@@ -2897,7 +2899,29 @@ BlockInputStreamPtr Segment::getBitmapFilterInputStream(const DMContext & dm_con
         real_ranges,
         filter ? filter->rs_operator : EMPTY_RS_OPERATOR,
         max_version,
-        expected_block_size);
+        read_data_block_rows);
+}
+
+size_t Segment::clipBlockRows(const Context & context, size_t expected_block_rows, const ColumnDefines & read_columns)
+{
+    size_t max_block_bytes = context.getSettingsRef().max_block_bytes;
+    size_t pack_rows = context.getSettingsRef().dt_segment_stable_pack_rows; // At least one pack.
+    return clipBlockRows(max_block_bytes, pack_rows, expected_block_rows, read_columns);
+}
+
+size_t Segment::clipBlockRows(size_t max_block_bytes, size_t pack_rows, size_t expected_block_rows, const ColumnDefines & read_columns)
+{
+    if (unlikely(max_block_bytes <= 0)) // Disable block bytes limit.
+    {
+        return expected_block_rows;
+    }
+    else
+    {
+        auto row_bytes = std::max(1, stable->avgRowBytes(read_columns)); // Avoid row_bytes to be 0.
+        auto rows = max_block_bytes / row_bytes;
+        rows = std::max(rows / pack_rows * pack_rows, pack_rows); // Align with pack and at least read one pack.
```

Comment:
```suggestion
        rows = std::ceil(rows / pack_rows) * pack_rows; // Align with pack and at least read one pack.
```
just a little suggestion